### PR TITLE
[move] Implement generic comparison method in move cmp::compare

### PR DESCRIPTION
**Pull request by igor-aptos**
## Description
Implement generic comparison operator for any move type. 
This enables writing generic methods that require comparison - for example binary_search on a vector.

Since move doesn't have signed integers, comparison return value is not intuitive (increased by 1 from usual -1, 0, 1), and so is wrapped into Ordering struct - so users don't need to think about it (with move 2, calling those will be pretty intuitive)

## Type of Change
- [x] New feature

## Which Components or Systems Does This Change Impact?
- [x] Move/Aptos Virtual Machine

## How Has This Been Tested?
Provided unit tests

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: aptos-move/framework/aptos-stdlib/sources/cmp.move
Comment (runtian-zhou) [inline aptos-move/framework/aptos-stdlib/sources/cmp.move:13]: Can we use enum for this now? cc @wrwg

File: aptos-move/framework/aptos-stdlib/sources/cmp.move
Comment (runtian-zhou) [inline aptos-move/framework/aptos-stdlib/sources/cmp.move:79]: Why is enum ignored?

Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/cmp.move:79]: enum's are not yet enabled in the framework.

File: aptos-move/framework/aptos-stdlib/sources/cmp.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/cmp.move:13]: would enum be better here? Would enums be more inefficient here? 

how would the native function return an enum that is defined in the module here?

File: third_party/move/move-vm/types/src/values/values_impl.rs
Comment (msmouse) [inline third_party/move/move-vm/types/src/values/values_impl.rs:590]: Shall we use `Display` instead of `Debug` in the message (hopefully that means more stable error message)?

File: third_party/move/move-vm/types/src/values/values_impl.rs
Comment (msmouse) [inline third_party/move/move-vm/types/src/values/values_impl.rs:688]: ditto

File: third_party/move/move-vm/types/src/values/values_impl.rs
Comment (igor-aptos) [reply third_party/move/move-vm/types/src/values/values_impl.rs:590]: this is copied from equals. I am not sure what is more stable from MoveVM prespective, but seems like this should be the same as in equals. 

and changing equals would require feature flag? if we wanted to change eeverything to {} , we should do that separately, I think

Comment (msmouse) [reply third_party/move/move-vm/types/src/values/values_impl.rs:590]: okay.. putting Debug text to messages we dare not to change generally make me very sad but you don't make it worse I guess

Comment (igor-aptos) [reply third_party/move/move-vm/types/src/values/values_impl.rs:590]: @vgao1996 for thoughts

File: aptos-move/aptos-gas-schedule/src/gas_schedule/move_stdlib.rs
Comment (lightmark) [inline aptos-move/aptos-gas-schedule/src/gas_schedule/move_stdlib.rs:43]: how do we get 367?

File: aptos-move/framework/move-stdlib/sources/cmp.move
Comment (lightmark) [inline aptos-move/framework/move-stdlib/sources/cmp.move:4]: why not enum?

File: aptos-move/framework/move-stdlib/sources/cmp.move
Comment (lightmark) [inline aptos-move/framework/move-stdlib/sources/cmp.move:15]: less_than, equal, greater_than can be ordering and return value.

**Comment by igor-aptos** [comment]
I cannot use move 2 features (i.e. enums), until framework is compiled with it. 

I can modify it in additional commit in the stack, and decide when landing.

**Comment by igor-aptos** [comment]
updated to use enums

File: aptos-move/aptos-gas-schedule/src/gas_schedule/move_stdlib.rs
Comment (igor-aptos) [reply aptos-move/aptos-gas-schedule/src/gas_schedule/move_stdlib.rs]: copied from equality costs

File: aptos-move/framework/aptos-stdlib/doc/cmp.md
Comment (wrwg) [inline aptos-move/framework/aptos-stdlib/doc/cmp.md:1]: That is surprising -- why do we have this .md in the aptos-stdlib? Leftover?

File: aptos-move/framework/move-stdlib/sources/cmp.move
Comment (wrwg) [inline aptos-move/framework/move-stdlib/sources/cmp.move:33]: s/__/_

File: aptos-move/framework/move-stdlib/sources/cmp.move
Comment (wrwg) [inline aptos-move/framework/move-stdlib/sources/cmp.move:15]: Please add a `///` doc comment to this function. You should likely outline that you do lexicographical ordering, with one or two examples.

File: aptos-move/framework/move-stdlib/sources/cmp.move
Comment (wrwg) [inline aptos-move/framework/move-stdlib/sources/cmp.move:45]: For good test coverage, a nested enum might be helpful, and have a field with a vector. Also strings and signers would be interesting.

File: aptos-move/framework/move-stdlib/sources/cmp.move
Comment (wrwg) [inline aptos-move/framework/move-stdlib/sources/cmp.move:76]: Test Equal case?

File: aptos-move/framework/move-stdlib/src/natives/cmp.rs
Comment (wrwg) [inline aptos-move/framework/move-stdlib/src/natives/cmp.rs:30]: Is that a left over C&P from the string module? Sounds familiar...

File: aptos-move/framework/move-stdlib/sources/cmp.move
Comment (wrwg) [inline aptos-move/framework/move-stdlib/sources/cmp.move:2]: Why those constants and enums?

File: aptos-move/framework/move-stdlib/src/natives/cmp.rs
Comment (wrwg) [inline aptos-move/framework/move-stdlib/src/natives/cmp.rs:37]: Forgot to update

File: aptos-move/framework/move-stdlib/src/natives/cmp.rs
Comment (wrwg) [inline aptos-move/framework/move-stdlib/src/natives/cmp.rs:39]: Probably then this one too.

File: third_party/move/move-vm/types/src/values/values_impl.rs
Comment (wrwg) [inline third_party/move/move-vm/types/src/values/values_impl.rs:570]: Does equality already do the same? 

I wonder whether this should use the same error code as equality, which definitely is not VM_EXTENSION__ERROR.

File: third_party/move/move-vm/types/src/values/values_impl.rs
Comment (wrwg) [inline third_party/move/move-vm/types/src/values/values_impl.rs:589]: The error message should not be 'cannot compare values' but more like 'inconsistent types for arguments of comparison" or similar. Should never happen with paranoid check, but nevertheless.

File: third_party/move/move-vm/types/src/values/values_impl.rs
Comment (wrwg) [reply third_party/move/move-vm/types/src/values/values_impl.rs:590]: The error message of VMError is (more or less unfortunately, making it hard to show to users) not stored on chain, so I do not think it is very important to be stable.

File: aptos-move/framework/move-stdlib/sources/cmp.move
Comment (ziaptos) [inline aptos-move/framework/move-stdlib/sources/cmp.move:93]: `GreaterThan` and `Equal` missing

File: aptos-move/framework/move-stdlib/sources/cmp.move
Comment (ziaptos) [reply aptos-move/framework/move-stdlib/sources/cmp.move:76]: Ditto for the tests below.

File: aptos-move/framework/move-stdlib/sources/cmp.move
Comment (ziaptos) [inline aptos-move/framework/move-stdlib/sources/cmp.move:8]: I'd strongly prefer to follow rust here, and have these be `Less`, `Greater` and `Equal`

File: aptos-move/framework/move-stdlib/sources/cmp.move
Comment (ziaptos) [reply aptos-move/framework/move-stdlib/sources/cmp.move:2]: Curious as well

File: third_party/move/move-vm/types/src/values/values_impl.rs
Comment (ziaptos) [reply third_party/move/move-vm/types/src/values/values_impl.rs:570]: Comment copy-pasted from fn equals. Let's at least change Disallow equality to Disallow comparison. I'd also remove rest of the comment and just write (see ValueImpl::fn equals).

@wrwg The VM_EXTENSION_ERROR makes sense, it's an error of the aggregators extension.

File: aptos-move/framework/aptos-stdlib/doc/cmp.md
Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/doc/cmp.md:1]: I've modeled by rust, just a longer version:

`is_lt` vs `is_le`

https://doc.rust-lang.org/std/cmp/enum.Ordering.html

@ziaptos given that, what do you prefer? Should I change to shorthands (i.e. same as rust), or to one of the longer variants?

Comment (ziaptos) [reply aptos-move/framework/aptos-stdlib/doc/cmp.md:1]: I'd prefer rust short ones.

**Comment by igor-aptos** [comment]
addressed comments and added tests

showed also that doing bcs::to_bytes and comparing would produce different and not natural ordering (due to little-endian serialization), showcasing the need for cmp::compare for primitive types as well, not just for performance.

**Comment by igor-aptos** [comment]
moved move files with enums into separate PR (#15245), so this can land without waiting on move2

**Review by runtian-zhou — CHANGES_REQUESTED** [review]
I would aim for 100% code coverage given the number of matches in the `value_impl.rs`. Should we land the move file first as a private function that is not called elsewhere just so we can add the tests first?

File: third_party/move/move-vm/types/src/values/values_impl.rs
Comment (runtian-zhou) [inline third_party/move/move-vm/types/src/values/values_impl.rs:819]: Was wondering if the logic would be easier if you do a read_ref on both ends and compare?

File: third_party/move/move-vm/types/src/values/values_impl.rs
Comment (runtian-zhou) [inline third_party/move/move-vm/types/src/values/values_impl.rs:565]: Can a `Container` be compared with `IndexedRef` etc?

**Comment by igor-aptos** [comment]
@runtian-zhou do you mean 100% code coverage with move unit tests, or something else?

For the complexity of code - code (i.e. all the matches) have been copied 1-1 from equals, so they should cover all the cases exactly.

signature of this move function uses enums, so I cannot land the move file with private function, until move 2 is enabled for the framework. So I've separated them out - but will confirm all tests in #15245 pass, before I land this one.

File: third_party/move/move-vm/types/src/values/values_impl.rs
Comment (igor-aptos) [reply third_party/move/move-vm/types/src/values/values_impl.rs:565]: no, as equals doesn't have those either.

**Review by runtian-zhou — APPROVED** [review]

**Review by ziaptos — APPROVED** [review]
I've reviewed this before, but didn't stamp

File: third_party/move/move-vm/types/src/values/values_impl.rs
Comment (igor-aptos) [reply third_party/move/move-vm/types/src/values/values_impl.rs:819]: not sure, why equals doesn't do that? maybe perf?

**Pull request merged**
